### PR TITLE
as: respect io.maps flags

### PR DESCRIPTION
**Pull request by jacobrosenthal**
make r_core_get_boundaries_prot respect io.maps flags. ie `/as @e:search.in=io.maps.exec` 

I think this is how it should work? I cargo culted what was being done for io.sections, but please evaluate.

Also I removed the duplicate r_core_get_boundaries_prot lookup in do_syscall_search the because cmd_search already does this and passes it in. In fact many of the commands in cmd_search are duplicating that lookup and really need to be evaluated. 

This significantly sped up my /as search which was hanging forever, presumably searching a bunch of register memory I have mapped that shouldnt be searched

### CODE REVIEW

**Comment by jacobrosenthal** [comment]
separate Issue to discuss other duplicate r_core_get_boundaries_prot https://github.com/radare/radare2/blob/master/libr/core/cmd_search.c#L3012

**Comment by sivaramaaa** [comment]
Nice work ! , but i guess this one is like a feature addition and not a bug fix , since we dont have `io.maps.exec` etc previously , but still i think this is usefull !

```
[0x00000000]> e anal.in=??
Valid values for search.in (depends on .from/.to and io.va):
raw               search in raw io (ignoring bounds)
block             search in the current block
io.map            search in current map
io.maps           search in all maps
io.section        search in current mapped section
io.sections       search in all mapped sections
io.sections.write search in all writable marked sections
io.sections.exec  search in all executable marked sections
dbg.stack         search in the stack
dbg.heap          search in the heap
dbg.map           search in current memory map
dbg.maps          search in all memory maps
dbg.maps.exec     search in all executable marked memory maps
dbg.maps.write    search in all writable marked memory maps
anal.fcn          search in the current function
anal.bb           search in the current basic-block
```

Related to : #10023

File: libr/core/cmd_search.c [hunk 1]
```diff
@@ -629,29 +629,50 @@ R_API RList *r_core_get_boundaries_prot(RCore *core, int protection, const char
 		if (m) {
 			append_bound (list, core->io, search_itv, m->itv.addr, m->itv.size);
 		}
-	} else if (!strcmp (mode, "io.maps")) { // Non-overlapping RIOMap parts not overriden by others (skyline)
+	} else if (!strncmp (mode, "io.maps", sizeof ("io.maps") - 1)) { // Non-overlapping RIOMap parts not overriden by others (skyline)
 		const RPVector *skyline = &core->io->map_skyline;
 		ut64 begin = UT64_MAX;
 		ut64 end = UT64_MAX;
 		size_t i;
+		int mask = 0;
```
Comment (MaskRay) [inline libr/core/cmd_search.c]: `mask = R_IO_WRITE | R_IO_EXEC;` and remove `!mask` below.

File: libr/core/cmd_search.c [hunk 2]
```diff
@@ -629,29 +629,50 @@ R_API RList *r_core_get_boundaries_prot(RCore *core, int protection, const char
...
+		bool readonly = false;
+
+		if (!strcmp (mode, "io.maps.exec")) {
```
Comment (MaskRay) [inline libr/core/cmd_search.c]: Can I express sections that are both executable and writable?

**Comment by Maijin** [comment]
This is https://github.com/radare/radare2/issues/10023

**Comment by MaskRay** [comment]
This might be a notation you don't like but I'd still like to propose it:

io.exec.w -> mask & R_IO_WRITE
io.exec.x -> mask & R_IO_EXEC
io.exec.wx -> mask & R_IO_WRITE && mask & R_IO_EXEC
io.exec.ro -> mask == R_IO_READ

File: libr/core/cmd_search.c [hunk 3]
```diff
@@ -629,29 +629,50 @@ R_API RList *r_core_get_boundaries_prot(RCore *core, int protection, const char
...
+			mask = R_IO_EXEC;
+		}
+		if (!strcmp (mode, "io.maps.write")) {
+			mask = R_IO_WRITE;
+		}
+		if (!strcmp (mode, "io.maps.readonly")) {
```
[the remark this answers is not in the document]
Comment (jacobrosenthal) [reply libr/core/cmd_search.c]: Beats me. I figured youd accept code that was already in upstream so this implementation stole heavily from io.sections here
https://github.com/radare/radare2/blob/master/libr/core/cmd_search.c#L713-L730

**Comment by jacobrosenthal** [comment]
@pancake Im not sure what action you want me to take from statement?
Delete the io.section.whatever code paths and remove it from help?

**Comment by radare** [comment]
and will be good to have bin.sections, too

File: libr/core/cmd_search.c (hunk 2 above)
Comment (radare) [reply libr/core/cmd_search.c]: i think this should go into a separate PR, but i agree that current approach have some limitations

**Comment by MaskRay** [comment]
@jacobrosenthal I was suggesting `io.sections.rw` ... because I think it provides more flexibility over `io.sections.exec` `io.sections.readonly`. If the notation is accepted, the other `search.in` choices can be updated separately.

**Comment by radare** [comment]
i think we can simplify that *.*.readonly, ... into something like *.*.[rwx] and assume that anything after io.maps. or bin.sections.* or dbg.maps.*, etc...  is a rwx string that must be parsed to determine the permissions, this way its cleaner and simpler

**Comment by radare** [comment]
oh well i did a poc on that and i found another problem in this. basically we need a way to specify if we want all pages with READ permission OR all pages that have ONLY the READ permission. (for example) how to express that without the *only thing? one option would be to also parse ".only" in that rwx string, and then use == isntead of & in the comparison. ill push my changes in a PR now, so you can see

**Comment by radare** [comment]
see https://github.com/radare/radare2/pull/11109

File: libr/core/cmd_search.c [hunk 4]
```diff
@@ -629,29 +629,50 @@ R_API RList *r_core_get_boundaries_prot(RCore *core, int protection, const char
...
+			mask |= R_IO_EXEC;
+		}
+		if (!strcmp (mode, "io.maps.write")) {
+			mask |= R_IO_WRITE;
+		}
+		if (!strcmp (mode, "io.maps.readonly")) {
+			readonly = true;
+		}
+
 		for (i = 0; i < r_pvector_len (skyline); i++) {
```
Comment (radare) [inline libr/core/cmd_search.c]: if skyline is going to die or change this code will need to be changed, im not sure whats the best way to walk into overlapped io maps. cc @MaskRay @condret

File: libr/core/cmd_search.c [hunk 5]
```diff
@@ -1764,7 +1784,8 @@ static void do_syscall_search(RCore *core, struct search_parameters *param) {
 			esp32 = r_str_newf ("%s,=", reg);
 		}
 	}
-	r_list_foreach (list, iter, map) {
+
+	r_list_foreach (param->boundaries, iter, map) {
```
Comment (radare) [inline libr/core/cmd_search.c]: i think this is the only change from your PR that must be merged. the other stuff is fixed in a better way in my pr

File: libr/core/cmd_search.c (hunk 1 above)
Comment (radare) [reply libr/core/cmd_search.c]: see my pr

File: libr/core/cmd_search.c (hunk 5 above)
Comment (jacobrosenthal) [reply libr/core/cmd_search.c]: Ill add this change to this other PR https://github.com/radare/radare2/pull/11103 and close this pr

File: libr/core/cmd_search.c (hunk 4 above)
Comment (MaskRay) [reply libr/core/cmd_search.c]: skyline and submaps are the same thing, sorted array of un-shadowed IO maps.

Their construction is different:

* skyline: O(n log n) rebuild but batch insertion is supported
* submaps: O(n) which also relies on r_oids* stuff. The overall complexity is higher (you also have to learn idpool stuff). The 

I don't see why "skyline is going to die". I have strong opinion regarding the implementation of that submaps PR.

**Pull request closed without merging**